url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockInnerTriangularSolver.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2009-2013, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.block;
+
+/**
+ * <p>
+ * Contains triangular solvers for inner blocks of a {@link org.ejml.data.BlockMatrix64F}.
+ * </p>
+ *
+ * <p>
+ * Algorithm for lower triangular inverse:<br>
+ *
+ * <pre>
+ * for i=1:m
+ *     for j=1:i-1
+ *         val = 0
+ *         for k=j:i-1
+ *             val = val - L(i,k) * X(k,j)
+ *         end
+ *         x(i,j) = val / L(i,i)
+ *     end
+ *     x(i,i) = 1 / L(i,i)
+ * end
+ * </pre> 
+ * </p>
+ *
+ * @author Peter Abeles
+ */
+public class BlockInnerTriangularSolver {
+
+    /**
+     * <p>
+     * Inverts a square lower triangular matrix:  L = L<sup>-1</sup>
+     * </p>
+     *
+     * @param L Lower triangular matrix being inverted. Not modified.
+     * @oaran K_inv Where the inverse is stored.  Can be the same as L.  Modified.
+     * @param m The number of rows and columns.
+     * @param offsetL which index does the L matrix start at.
+     * @param offsetL_inv which index does the L_inv matrix start at.
+     *
+     */
+    public static void invertLower( double L[] ,
+                                    double L_inv[] ,
+                                    int m ,
+                                    int offsetL ,
+                                    int offsetL_inv )
+    {
+        for( int i = 0; i < m; i++ ) {
+            double L_ii = L[ offsetL + i*m + i ];
+            for( int j = 0; j < i; j++ ) {
+                double val = 0;
+                for( int k = j; k < i; k++ ) {
+                    val += L[offsetL + i*m + k] * L_inv[ offsetL_inv + k*m + j ];
+                }
+                L_inv[ offsetL_inv + i*m + j ] = -val / L_ii;
+            }
+            L_inv[ offsetL_inv + i*m + i ] =  1.0 / L_ii;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 65:75, 92:102.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.